src/parquet/arrow/writer.cc

```diff
@@ -430,18 +430,17 @@ Status FileWriter::Impl::TypedWriteBatch<ByteArrayType, ::arrow::BinaryType>(
 
   if (writer->descr()->schema_node()->is_required() || (data->null_count() == 0)) {
     // no nulls, just dump the data
-    for (int64_t i = 0; i < num_values; i++) {
-      buffer_ptr[i] =
-          ByteArray(data->value_length(i + offset), data_ptr + data->value_offset(i));
+    for (int64_t i = 0; i < data->length(); i++) {
+      buffer_ptr[i] = ByteArray(data->value_length(i), data_ptr + data->value_offset(i));
```

Comment:
For performance, recommend caching `const int32_t* value_offsets = data->raw_value_offsets()` (which is shifted) and using the offsets to compute the value lengths